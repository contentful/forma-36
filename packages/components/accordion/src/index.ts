export { Accordion } from './Accordion';
export type { AccordionProps } from './Accordion';
export { AccordionHeader } from './AccordionHeader/AccordionHeader';
export type { AccordionHeaderProps } from './AccordionHeader/AccordionHeader';
export { AccordionItem } from './AccordionItem/AccordionItem';
export type { AccordionItemProps } from './AccordionItem/AccordionItem';
export { AccordionPanel } from './AccordionPanel/AccordionPanel';
export type { AccordionPanelProps } from './AccordionPanel/AccordionPanel';
