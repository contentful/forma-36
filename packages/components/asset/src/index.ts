export { Asset } from './Asset';
export type { AssetProps } from './Asset';
