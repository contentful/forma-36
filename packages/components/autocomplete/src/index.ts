export { Autocomplete } from './Autocomplete';
export type { AutocompleteProps } from './Autocomplete';
