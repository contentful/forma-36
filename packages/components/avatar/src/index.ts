export { Avatar } from './Avatar';
export type { AvatarProps } from './Avatar';
export { AvatarGroup } from './AvatarGroup';
export type { AvatarGroupProps } from './AvatarGroup';
