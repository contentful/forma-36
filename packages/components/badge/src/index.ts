export { Badge } from './Badge';
export type { BadgeProps } from './Badge';
export { EntityStatusBadge } from './EntityStatusBadge';
export type { EntityStatusBadgeProps } from './EntityStatusBadge';
export type { BadgeVariant } from './types';
