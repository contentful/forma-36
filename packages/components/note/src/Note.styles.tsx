import { css } from 'emotion';
import type { CSSObject } from '@emotion/serialize';
import tokens from '@contentful/f36-tokens';

import type { NoteProps, NoteVariant } from './Note';

const variantToStyles = (variant: NoteVariant): CSSObject => {
  switch (variant) {
    case 'primary':
      return {
        backgroundColor: tokens.blue100,
        borderColor: tokens.blue300,

        a: {
          color: tokens.blue700,
        },
      };
    case 'positive':
      return {
        backgroundColor: tokens.green100,
        borderColor: tokens.green300,

        a: {
          color: tokens.blue700,
        },
      };
    case 'negative':
      return {
        backgroundColor: tokens.red100,
        borderColor: tokens.red300,

        a: {
          color: tokens.blue700,
        },
      };
    case 'warning':
      return {
        backgroundColor: tokens.orange100,
        borderColor: tokens.orange300,

        a: {
          color: tokens.blue700,
        },
      };
    case 'neutral':
      return {
        backgroundColor: tokens.gray100,
        borderColor: tokens.gray300,

        a: {
          color: tokens.blue700,
        },
      };
    default:
      return {};
  }
};

export const getNoteStyles = () => {
  return {
    container: ({ variant }: Pick<NoteProps, 'variant'>) =>
      css({
        position: 'relative',
        borderRadius: tokens.borderRadiusMedium,
        border: '1px solid',
        ...variantToStyles(variant),
      }),
    title: css({
      color: tokens.gray800,
      fontSize: tokens.fontSizeL,
      lineHeight: tokens.lineHeightL,
    }),
    description: css({
      color: tokens.gray700,
    }),
    close: css({
      position: 'absolute',
      top: tokens.spacingXs,
      right: tokens.spacingXs,
      padding: tokens.spacingXs,
      '&:hover, &:active': {
        backgroundColor: 'transparent',
      },
    }),
    closeIcon: css({
      fill: tokens.gray600,
    }),
  };
};
