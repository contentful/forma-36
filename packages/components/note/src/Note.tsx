import { cx } from 'emotion';
import React from 'react';
import { Flex, Grid } from '@contentful/f36-core';
import type {
  CommonProps,
  PropsWithHTMLElement,
  ExpandProps,
} from '@contentful/f36-core';
import { Button } from '@contentful/f36-button';
import { Heading, Paragraph, Text } from '@contentful/f36-typography';
import {
  CheckCircleIcon,
  CloseIcon,
  ErrorCircleIcon,
  InfoCircleIcon,
  WarningIcon,
} from '@contentful/f36-icons';
import { Icon } from '@contentful/f36-icon';

import { getNoteStyles } from './Note.styles';

const icons = {
  primary: InfoCircleIcon,
  positive: CheckCircleIcon,
  negative: ErrorCircleIcon,
  warning: WarningIcon,
  neutral: InfoCircleIcon,
};

export type NoteVariant =
  | 'negative'
  | 'positive'
  | 'primary'
  | 'warning'
  | 'neutral';

export type NoteInternalProps = CommonProps & {
  /**
   * Determines style variation of Note component
   */
  variant?: NoteVariant;
  /**
   * Sets title in the Note
   */
  title?: React.ReactNode;
  /**
   * Children of Note
   */
  children?: React.ReactNode | string;
  /**
   * Defines if the close button should be rendered
   * @default false
   */
  withCloseButton?: boolean;
  /**
   * Callback for handling closing
   */
  onClose?: React.MouseEventHandler<HTMLButtonElement>;
};
export type NoteProps = PropsWithHTMLElement<NoteInternalProps, 'article'>;

/**
 * @description: Note provides context and information about a situation or action.
 */
export const Note = React.forwardRef<HTMLElement, ExpandProps<NoteProps>>(
  (props, ref) => {
    const {
      children,
      className,
      withCloseButton = false,
      variant = 'primary',
      onClose,
      testId = 'cf-ui-note',
      title,
      ...otherProps
    } = props;

    const styles = getNoteStyles();

    return (
      <Grid
        {...otherProps}
        columns={withCloseButton ? 'auto 1fr 24px' : 'auto 1fr'} // 24px is the width of the close button
        as="article"
        className={cx(styles.container({ variant }), className)}
        testId={testId}
        ref={ref}
        padding="spacingM"
      >
        <Icon
          as={icons[variant]}
          variant={variant === 'neutral' ? 'muted' : variant}
          size={title ? 'medium' : 'small'}
        />
        <Flex flexDirection="column">
          {title && (
            <Heading
              as="h2"
              className={styles.title}
              marginBottom={!children ? 'none' : 'spacingS'}
            >
              {title}
            </Heading>
          )}
          {children && (
            <Text
              as="div"
              lineHeight="lineHeightM"
              className={styles.description}
            >
              {typeof children === 'string' ? (
                <Paragraph marginBottom="none">{children}</Paragraph>
              ) : (
                children
              )}
            </Text>
          )}
        </Flex>
        {withCloseButton && (
          <Button
            variant="transparent"
            startIcon={<CloseIcon className={styles.closeIcon} />}
            onClick={onClose}
            testId={`${testId}-close`}
            aria-label="Dismiss"
            className={styles.close}
          />
        )}
      </Grid>
    );
  },
);

Note.displayName = 'Note';
