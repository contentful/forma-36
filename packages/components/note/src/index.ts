export { Note } from './Note';
export type { NoteProps } from './Note';
