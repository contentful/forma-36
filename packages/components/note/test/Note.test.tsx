import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Note } from '../src/Note';
import { axe } from '@/scripts/test/axeHelper';

describe('Note', function () {
  const noteText =
    'A piece of information that is relevant to the context the user is currently in.';
  it('renders the component', () => {
    const { container } = render(<Note>{noteText}</Note>);

    expect(container.firstChild).toBeTruthy();
  });

  it('renders the component with title', () => {
    const { container, getByText } = render(
      <Note variant="positive" title="Positive title">
        {noteText}
      </Note>,
    );

    const title = getByText('Positive title');
    expect(title).toBeTruthy();
    expect(container.firstChild).toContainElement(title);
  });

  it('renders the component with an additional class name', () => {
    const { container } = render(
      <Note className="my-extra-class">{noteText}</Note>,
    );

    expect(container.firstChild).toHaveClass('my-extra-class');
  });

  it('renders the component with a close button', () => {
    const mockOnClose = jest.fn();
    const { container, getByRole } = render(
      <Note hasCloseButton onClose={mockOnClose}>
        {noteText}
      </Note>,
    );

    const button = getByRole('button');
    expect(container.firstChild).toContainElement(button);
    fireEvent.click(button);
    expect(mockOnClose).toHaveBeenCalled();
  });

  it(`has no a11y issues`, async () => {
    const { container } = render(<Note>{noteText}</Note>);

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
