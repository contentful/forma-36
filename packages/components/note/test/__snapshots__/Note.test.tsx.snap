// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Note renders the component 1`] = `
.emotion-4 {
  padding: 1rem;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 400;
  line-height: 1.5;
  font-size: 0.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  border-radius: 6px;
  padding-right: 1rem;
  border: 1px solid;
  background-color: #E8F5FF;
  border-color: #98CBFF;
}

.emotion-4 a {
  color: #0041AB;
}

.emotion-3 {
  margin-left: 0.5rem;
  fill: currentColor;
  margin-right: unset;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0.75rem;
}

.emotion-0 {
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
}

.emotion-2 {
  display: block;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 600;
  color: #111B2B;
  font-size: 1.25rem;
  line-height: 2rem;
  text-rendering: optimizeLegibility;
  font-weight: 400;
  color: #414D63;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #414D63;
  width: 100%;
}

<article
  class="emotion-4"
  data-test-id="cf-ui-note"
>
  <div
    class="emotion-1"
  >
    <svg
      aria-hidden="true"
      class="emotion-0"
      data-test-id="cf-ui-icon"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
  </div>
  <div
    class="emotion-3"
  >
    <p
      class="emotion-2"
      data-test-id="cf-ui-text"
    >
      A piece of information that is relevant to the context the user is currently in.
    </p>
  </div>
</article>
`;

exports[`Note renders the component with an additional class name 1`] = `
.emotion-4 {
  padding: 1rem;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 400;
  line-height: 1.5;
  font-size: 0.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  border-radius: 6px;
  padding-right: 1rem;
  border: 1px solid;
  background-color: #E8F5FF;
  border-color: #98CBFF;
}

.emotion-4 a {
  color: #0041AB;
}

.emotion-3 {
  margin-left: 0.5rem;
  fill: currentColor;
  margin-right: unset;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0.75rem;
}

.emotion-0 {
  display: inline-block;
  fill: #0059C8;
  height: 18px;
  width: 18px;
}

.emotion-2 {
  display: block;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 600;
  color: #111B2B;
  font-size: 1.25rem;
  line-height: 2rem;
  text-rendering: optimizeLegibility;
  font-weight: 400;
  color: #414D63;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #414D63;
  width: 100%;
}

<article
  class="my-extra-class emotion-4"
  data-test-id="cf-ui-note"
>
  <div
    class="emotion-1"
  >
    <svg
      aria-hidden="true"
      class="emotion-0"
      data-test-id="cf-ui-icon"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
      />
    </svg>
  </div>
  <div
    class="emotion-3"
  >
    <p
      class="emotion-2"
      data-test-id="cf-ui-text"
    >
      A piece of information that is relevant to the context the user is currently in.
    </p>
  </div>
</article>
`;

exports[`Note renders the component with title 1`] = `
.emotion-4 {
  margin-left: 0.5rem;
  fill: currentColor;
  margin-right: unset;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 0.75rem;
}

.emotion-3 {
  display: block;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 600;
  color: #111B2B;
  font-size: 1.25rem;
  line-height: 2rem;
  text-rendering: optimizeLegibility;
  font-weight: 400;
  color: #414D63;
  font-size: 0.875rem;
  line-height: 1.25rem;
  color: #414D63;
  width: 100%;
}

.emotion-5 {
  padding: 1rem;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 400;
  line-height: 1.5;
  font-size: 0.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  border-radius: 6px;
  padding-right: 1rem;
  border: 1px solid;
  background-color: #EAF9E8;
  border-color: #9ED696;
}

.emotion-5 a {
  color: #00550C;
}

.emotion-0 {
  display: inline-block;
  fill: #008539;
  height: 24px;
  width: 24px;
}

.emotion-2 {
  display: block;
  margin: 0;
  font-family: -apple-system,BlinkMacSystemFont,Segoe UI,Helvetica,Arial,sans-serif,Apple Color Emoji,Segoe UI Emoji,Segoe UI Symbol;
  font-weight: 600;
  color: #111B2B;
  font-size: 1.25rem;
  line-height: 2rem;
  text-rendering: optimizeLegibility;
  color: #1B273A;
  font-weight: 500;
  font-size: 1rem;
  margin-bottom: 0.5rem;
}

<article
  class="emotion-5"
  data-test-id="cf-ui-note"
>
  <div
    class="emotion-1"
  >
    <svg
      aria-hidden="true"
      class="emotion-0"
      data-test-id="cf-ui-icon"
      viewBox="0 0 24 24"
    >
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
      <path
        d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
      />
    </svg>
  </div>
  <div
    class="emotion-4"
  >
    <h2
      class="emotion-2"
      data-test-id="cf-ui-heading"
    >
      Positive title
    </h2>
    <p
      class="emotion-3"
      data-test-id="cf-ui-text"
    >
      A piece of information that is relevant to the context the user is currently in.
    </p>
  </div>
</article>
`;
